core/src/main/scala/kafka/utils/ZkUtils.scala

```diff
@@ -892,3 +871,219 @@ object ZkPath {
+                  null)
+    } else {
+      zkHandle.create(prefix,
+                  new Array[Byte](0),
+                  Ids.OPEN_ACL_UNSAFE,
+                  CreateMode.PERSISTENT,
+                  new StringCallback() {
+                        def processResult(rc : Int,
+                                          path : String,
+                                          ctx : Object,
+                                          name : String) {
+                          Code.get(rc) match {
+                            case Code.OK => {
+                              // Nothing to do
+                            }
+                            case Code.NODEEXISTS => {
+                              // Nothing to do
+                            }
+                            case Code.CONNECTIONLOSS => {
+                              // try again
+                              val suffix = ctx.asInstanceOf[String]
+                              createRecursive(path, suffix)
+                            }
+                            case Code.NONODE => {
+                              error("No node for path %s (could be the parent missing)".format(path))
+                              setResult(Code.get(rc))
+                            }
+                            case Code.SESSIONEXPIRED => {
+                              error("Session has expired while creating %s".format(path))
+                              setResult(Code.get(rc))
+                            }
+                            case _ => {
+                              info("ZooKeeper event while creating registration node %s %s".format(path, Code.get(rc)))
+                              setResult(Code.get(rc))
+                            }
+                          }
+                        }
+                  },
+                  suffix)
+      // Update prefix and suffix
+      val index = suffix.indexOf('/', 1) match {
+        case -1 => suffix.length
+        case x : Int => x
+      }
+      // Get new prefix
+      val newPrefix = prefix + suffix.substring(0, index)
+      // Get new suffix
+      val newSuffix = suffix.substring(index, suffix.length)
+      createRecursive(newPrefix, newSuffix)
+    }
+  }
+
+  private def setResult(code: Code) {
+    result = code
+    latch.countDown()
+  }
+
+  private def waitUntilResolved(): Code = {
+    latch.await()
+    result
+  }
+
+  def createAndWatch() {
+    val index = path.indexOf('/', 1) match {
+        case -1 => path.length
+        case x : Int => x
+    }
+    val prefix = path.substring(0, index)
+    val suffix = path.substring(index, path.length)
+    debug("Path: %s, Prefix: %s, Suffix: %s".format(path, prefix, suffix))
+    createRecursive(prefix, suffix)
+    val result = waitUntilResolved()
+    result match {
+      case Code.OK => {
+        // Nothing to do
+      }
+      case _ => {
+        throw ZkException.create(KeeperException.create(result))
```

Comment:
In this case, caller of createAndWatch needs to handle ZkException for NONODE, NODEEXISTS, SESSIONEXPIRED, right? Could you add that in Controller / Broker / Consumer?